lib/dialects/oracle/connection-manager.js

```diff
@@ -0,0 +1,168 @@
+      
+      let connectString = config.host;
+
+      if (config.port && config.port !== 0) {
+        connectString += `:${config.port}`;
+      } else {
+        connectString += ':1521'; //Default port number
+      }
+      connectString += `/${config.database}`;
+      config.database = connectString;
+    }
+  }
+
+  // Expose this as a method so that the parsing may be updated when the user has added additional, custom types
+  _refreshTypeParser(dataType) {
+    parserStore.refresh(dataType);
+  }
+
+  _clearTypeParser() {
+    parserStore.clear();
+  }
+
+  connect(config) {
+    const connectionConfig = {
+      user: config.username,
+      host:  config.host,
+      port : config.port,
+      database : config.database,
+      password: config.password,
+      externalAuth : config.externalAuth,
+      stmtCacheSize : 0
+    };
+
+    //We check if there dialect options
+    if ('dialectOptions' in config) {
+      const dialectOptions = config.dialectOptions;
+
+      //If stmtCacheSize is defined, we set it
+      if (dialectOptions && 'stmtCacheSize' in dialectOptions) {
+        connectionConfig.stmtCacheSize = dialectOptions.stmtCacheSize;
+      }
+    }
+
+
+    //Check the config object
+    this.checkConfigObject(connectionConfig);
+
+    //We assume that the database has been correctly formed
+    connectionConfig.connectString = connectionConfig.database;
+
+    if (config.dialectOptions) {
+      Object.keys(config.dialectOptions).forEach(key => {
+        connectionConfig[key] = config.dialectOptions[key];
+      });
+    }
+
+    return this.lib.getConnection(connectionConfig).then(connection => {
+      return connection;
+    })
+      .catch(err => {
+        //We split to get the error number; it comes as ORA-XXXXX: 
+        let errorCode = err.message.split(':');
+        errorCode = errorCode[0];
+
+        if (errorCode) {
+          switch (errorCode) {
+            case 'ORA-28000': //Account locked
+              throw new sequelizeErrors.ConnectionRefusedError(err);
+            case 'ORA-01017': //ORA-01017 : invalid username/password; logon denied
+              throw new sequelizeErrors.AccessDeniedError(err);
+            case 'ORA-12154':
+              throw new sequelizeErrors.HostNotReachableError(err); //ORA-12154: TNS:could not resolve the connect identifier specified
+            case 'ORA-12514' : // ORA-12514: TNS:listener does not currently know of service requested in connect descriptor
+              throw new sequelizeErrors.HostNotFoundError(err);
+            case 'ORA-12541' : //ORA-12541: TNS:No listener
+              throw new sequelizeErrors.AccessDeniedError(err); 
+            default:
+              throw new sequelizeErrors.ConnectionError(err);
+          }
+        } else {
+          throw new sequelizeErrors.ConnectionError(err);
+        }
+
+        throw err; //Unknown error, we throw it anyway
+      });
+  }
+
+  disconnect(connection) {
```

Comment:
This doesn't look right. `disconnect` seems to want a promise that's resolved or rejected. The current implementation throws on the main thread. 

I suggest requiring in Utils at the top:
```javascript
const Utils = require('../../utils');
```

Then update disconnect to something like this:
```javascript
disconnect(connection) {
  return new Utils.Promise((resolve, reject) => {
    connection.release()
      .then(resolve)
      .catch(err => {
        reject(new sequelizeErrors.ConnectionError(err));
      });
  });
}
```

>